### ISSUE TITLE
KIWI assumes rpmdb is BDB (rpmdb can be different types of databases)

### ISSUE THREAD

**Issue by Conan-Kudo** [SUSE/kiwi#546]
## Problem description
Since RPM 4.13, the RPM database has been able to be stored in multiple database formats. As of RPM 4.14, there are three database formats: BDB, LMDB, and NDB.

However, there are places in the KIWI code that assume that RPM always is a Berkeley database, which this is not guaranteed. Thus, rpmdb-centric actions and manipulation need to account for all three database backends.

For example, in the Zypper package manager code: https://github.com/SUSE/kiwi/blob/master/kiwi/package_manager/zypper.py#L270-L285

The code linked above assumes a Berkeley database structure, assumes BDB tools will work on the data, and that everything will be sane afterward.

**Comment by schaefi** [comment]
You are right so far I was under the impression all linux distros of interest uses BDB for rpm as database backend and so kiwi's dump/reload code is based on that

Are you aware of any major distribution which have plans to move to another database backend ?
I'm asking because that would increase the priority of this issue

Thanks

**Comment by Conan-Kudo** [comment]
Fedora is considering LMDB, which is why I bring this up now. We have [a Fedora bug](https://bugzilla.redhat.com/show_bug.cgi?id=1086784) about it, and we're actively trying to make the stack happy with alternative backends (see https://github.com/openSUSE/libsolv/pull/123 and https://github.com/rpm-software-management/libdnf/issues/362).

**Comment by Conan-Kudo** [comment]
It's now possible to use libsolv with non-BDB rpmdb: https://github.com/openSUSE/libsolv/commit/ae214a654c4c0f3b43ca763e70bde4793d00422a

**Comment by davidcassany** [comment]
> It's now possible to use libsolv with non-BDB rpmdb: openSUSE/libsolv@ae214a6

Interesting, then I guess we should start considering some data base type detection strategy. Thanks for the update.

I don't think we are yet on hurry to solve that, but if we can start to make some tests in Tumbleweed I guess the issue can be considered to be in a workable state.
